src/stirling/source_connectors/socket_tracer/bcc_bpf/grpc_c_trace.c

```diff
@@ -0,0 +1,1557 @@
+    {
+        return -1;
+    }
+
+    switch (version)
+    {
+        case GRPC_C_V1_19_0:
+            *flow_controlled_buffer = stream + 0x6e8; // Found with IDA, after "sending trailing_metadata" string.
+            break;
+        case GRPC_C_V1_24_1:
+            *flow_controlled_buffer = stream + 0x980;
+            break;
+        case GRPC_C_V1_33_2:
+            *flow_controlled_buffer = stream + 0x970;
+            break;
+        case GRPC_C_V1_41_1:
+            *flow_controlled_buffer = stream + 0x978;
+            break;
+        default:
+            return -1;
+    }
+    return 0;
+}
+
+/*
+ * @brief   Get the slices of data inside a grpc_slice_buffer.
+ *
+ * @param   slice_buffer    The buffer of slices whose data should be extracted.
+ * @param   write_event_data
+ *                          The event data that will be used to fire perf events.
+ *                          For each slice, the data will be put here and fired
+ *                          to user-mode.
+ * @param   connection_info
+ *                          Data that relates to the HTTP2 connection on which the
+ *                          slices are sent. This is used to determine and update
+ *                          the position of the slice in the stream.
+ *
+ * @remark  The same struct is used for multiple perf events. This seems to work
+ *          fine, but we should keep an eye out for erros caused by this.
+ * @remark  The slice fired to user-mode is not "cleaned" (set to zeros) between
+ *          events. The user should use the slice length to determine the size,
+ *          because later indices may contain data from previous events.
+ *
+ * @return  -1 on failure.
+ *          0 on success.
+ */
+static inline u32 get_slices_from_grpc_slice_buffer_and_fire_perf_event_per_slice(
+    grpc_slice_buffer * slice_buffer,
+    struct grpc_c_event_data_t * write_event_data,
+    struct conn_info_t * connection_info,
+    struct pt_regs * ctx)
+{
+    struct grpc_c_data_slice_t * data_slice = NULL;
+    grpc_slice * slice = NULL;
+    u32 slice_length = 0;
+    void * slice_bytes = NULL;
+    u32 length_to_read = 0;
+    uint32_t amount_of_slices = 0;
+    uint32_t event_data_length = 0;
+
+    if (NULL == slice_buffer || NULL == write_event_data || NULL == connection_info || NULL == ctx)
+    {
+        return -1;
+    }
+
+    // Read amount of slices.
+    if (0 != bpf_probe_read(
+            &amount_of_slices,
+            sizeof(amount_of_slices),
+            (void *)(slice_buffer + 0x10)))
+    {
+        return -1;
+    }
+
+    // Read pointer to first slice.
+    if (0 != bpf_probe_read(
+            &slice,
+            sizeof(slice),
+            (void *)(slice_buffer + 0x08)))
+    {
+        return -1;
+    }
+
+    // If there are too many slices, only read the amount we're allowed to.
+#pragma unroll
+    for (u32 i = 0 ; i < SIZE_OF_DATA_SLICE_ARRAY ; i++)
+    {
+        if (i >= amount_of_slices)
```

Comment:
I generally tried to have the for condition as simple as possible so that the verifier sees that it's a bounded loop. But you're right, given the order of conditions is correct, there's no real reason not to.